### DISCUSSION TITLE
Validation patterns

### DISCUSSION THREAD

**Discussion by michaltaberski** [pocketbase/pocketbase#1483] [Q&A] [answered]
I have a simple structure with **1..*** relationship:

```jsx
type Event = {
  title: string,
  date: string, // date string
  organizer: string, // participantId
  participants: string[], // id[]
  participantsLimit: number,
}

type Participant = {
  name: string,
}
```

I would like to create a robust `joinEvent` function

```jsx
export const joinEvent = async (eventId: string, participantId: string) => {
  const event = await pb.collection('events').getOne(eventId);
  if (event.participantsLimit === event.participants.length) {
    throw new Error('Participants number exceeded')
  }
  return await pb.collection("events").update(eventId, {
    participants: [...event.participants, participantId],
  });
};
```

This can work, but it only works if I expose an entire record update.

1) How can I set validation on the PB level (backend) so only creator can update the event, but any user can join the event? (joining an event is also technically an event update 🤔 ).
2) Is it possible to apply the participants lenght validation on the backend?

#### // Edit

Perhaps the data sturcture should be changed?

```jsx
type Event = {
  title: string,
  date: string, // date string
  organizer: string, // participantId
  participantsLimit: number,
}

type Participant = {
  name: string,
  events: string[] // id[]
}
```

but that seems to me not natural. It would help with update validation, but validating the participants count would be a nightmare I'm afraid.

**Comment by ganigeorgiev** [comment, marked as answer]
> 1. How can I set validation on the PB level (backend) so only creator can update the event, but any user can join the event?. 

For you first example, you can do that with exclusion/direct-match of every other field. eg:
```js
@request.auth.id = organizer || (
  @request.auth.id != "" &&
  title = @request.data.title &&
  date = @request.data.date &&
  organizer = @request.data.organizer &&
  participantsLimit = @request.data.participantsLimit
)
```
The above will require to send the original `title`, `organizer`, `participantsLimit` with each `participants` update request.
_**In the upcoming v0.11.0 this no longer will be necessary and would be simplified with the `@request.data.*.isset/@isset` filter** (I'm still not sure about the modifier naming)._

But even if you decide to go with this approach, there is one major flow which is that it will allow one participant to remove/add other participants, and this may not be wanted depending on your use case.

Your second example, with the multi-`relation` field in the `participants` collection, solves this, but as you noted,
it complicates the client-side counts fetch.

A 3rd more traditional SQL approach, but may require a little more boilerplate, would be to use a junction collection, eg. `eventParticipant`, that could have only 2 single `relation` fields pointing to the `participants` and the `events` collections.
In this case you don't have to allow the `participants` update access to your `events` collection and can count client-side the participants by querying the `eventParticipant` collection. eg:

```js
const { totalItems } = await pb.collection('eventParticipant').getList(1, 1, { filter: `event='${eventId}'` });
```

If you want to fetch the participants with the event(s) in a single query, you can use [indirect expansion](https://pocketbase.io/docs/expanding-relations/#indirect-expand):

```js
const event = await pb.collection('events').getOne('EVENT_ID', { expand: 'eventParticipant(event).participant' });
```

_Personally, I would go with the junction collection option._

---

> 2. Is it possible to apply the participants lenght validation on the backend?

Unfortunately at the moment this can't be done in an API rule.
I'll consider in the upcoming v0.11/v0.12 to eventually add `.length/.@length` modifier, but for now if you want to validate the limit server-side you'll have to [use PocketBase as framework](https://pocketbase.io/docs/use-as-framework/) and bind to the [`OnRecordBeforeUpdateRequest` hook](https://pocketbase.io/docs/event-hooks/#record-api-hooks).
